src/components/Topology/Topology.js

```diff
@@ -240,25 +320,158 @@ export default ({ modelData, width, height }) => {
           : "circle(55px at 63px 63px)"
       );
 
-    const relationLines = topo.selectAll(".relation").data(relations);
-    const relationLine = relationLines.enter().insert("g", ":first-child");
+    function dragstarted() {
+      d3.select(this).select("circle").attr("stroke", "#E9531F");
+    }
+
+    function drag() {
+      const app = d3.select("circle", this);
+      const radius = app.attr("r");
+
+      d3.select(this).attr("transform", () => {
+        const iconX = d3.event.x - radius;
+        const iconY = d3.event.y - radius;
+
+        return `translate(${iconX}, ${iconY})`;
+      });
+      updateRelations(relationLine);
+    }
 
-    relationLine
-      .classed("relation", true)
-      .append("line")
-      .attr("x1", (d) => getRelationPosition(d).x1)
-      .attr("y1", (d) => getRelationPosition(d).y1)
-      .attr("x2", (d) => getRelationPosition(d).x2)
-      .attr("y2", (d) => getRelationPosition(d).y2)
-      .attr("stroke", "#666666")
-      .attr("stroke-width", 2);
+    function dragended() {
+      const app = d3.select(this);
+      app.select("circle").attr("stroke", "#888888");
+      // const name = app.attr("data-name");
+      // const annotations = getRect(app);
+      // const updateResponse = updateAnnotations(conn, name, annotations);
+      // updateResponse.then((data) => {
+      //   if (!data) {
+      //     setFlashMessage({
+      //       type: "error",
+      //       message: "Sorry - the position of your apps could not be updated.",
+      //     });
+      //   } else {
+      //     setFlashMessage({
+      //       type: "success",
+      //       message: "The position of your apps have been updated.",
+      //     });
+      //   }
+      // });
+    }
+
+    if (!isReadOnly) {
+      appIcon.call(
+        d3.drag().on("start", dragstarted).on("drag", drag).on("end", dragended)
+      );
+    }
+
+    setRelations(relationLine);
 
     appIcons.exit().remove();
     relationLines.exit().remove();
 
     return () => {
       topo.remove();
     };
-  }, [applications, deltaX, deltaY, height, width, maxX, maxY, relations]);
-  return <svg ref={ref} />;
+  }, [
+    applications,
+    conn,
```

Comment:
You shouldn't rely on conn, it has a timestamp in it so it'll change every time a request is ever transferred over it.